python/ray/serve/tests/test_deploy.py

```diff
@@ -337,7 +336,7 @@ async def __call__(self, request):
 
     # ref2 will block until the signal is sent.
     ref2 = call.remote(block=True)
-    assert len(ray.wait([ref2], timeout=0.1)[0]) == 0
+    assert len(ray.wait([ref2], timeout=2.1)[0]) == 0
```

Comment:
I have bumped the timeout here because I think the following tests execute before `call.remote` reaches `requests.get` (https://github.com/ray-project/ray/blob/e45383793f5608d7b9b04ce08a2406b27d9dd83b/python/ray/serve/tests/test_deploy.py#L304). Since the URL for both the deployments is the same so may be things change there by the time `requests.get` is even executed. cc: @pcmoritz @simon-mo 

https://github.com/ray-project/ray/blob/e45383793f5608d7b9b04ce08a2406b27d9dd83b/python/ray/serve/tests/test_deploy.py#L358

https://github.com/ray-project/ray/blob/e45383793f5608d7b9b04ce08a2406b27d9dd83b/python/ray/serve/tests/test_deploy.py#L370

Reply:
I checked. It seems like a race condition. When we use `0.1` as the timeout in `ray.wait` then, `V2.handler` is executed before `V1.handler` and hence `requests.get` receives `2|PID` as the value. Basically the version update on Windows happens before `request.get` completes in case of `ref2`. Increasing `timeout` to `2.1` gives more time for completing execution of `requests.get` and hence this race condition is avoided. cc: @pcmoritz @simon-mo Let me know if anything is not clear.

P.S. I am not sure if we can make this stuff more safer because this condition can arise on any OS depending on the speed of `requests.get`. May be we can make the tests more safer like verifying the version in `def call(version, block=False)` before going for `requests.get`. execution.

Reply:
The `assert` statement is checking whether `ref2` is pointing to a remote function that has finished executing. Since `block=True` in `ref2`'s definition, and since both the `ray.get` and `requests.get` calls are blocking operations, `call.remote` should never be ready before the `assert` statement. This is why the `assert` checks whether the list of references that are ready is 0.

I'm probably missing something, but I don't think there should be a race condition here. The `assert` statement should pass before `V2` is deployed, which happens after the assertion. Was the `assert` statement failing without bumping the timeout? Do you happen to know what the `ray.wait` call's value was?

Reply:
The test which fails without timeout bump here is https://github.com/ray-project/ray/blob/76b91efd9b4569c81abfdd67ac88675f8c58ef02/python/ray/serve/tests/test_deploy.py#L369 and not the above one.

Following are the points that I want to highlight, 

1. The test in L369 fails only when `use_handle = False`.
2. `val2` is `"2"` and hence the failure.
3. This means that when `requests.get` got executed and accessed the location passed to it, `V2` was already deployed on that location and hence `"2"` as the output of `val2`.
4. Adding a timeout of 2 extra seconds gives `val2` as `"1"`.
5. Infact, I didn’t touch the timeout and added a `time.sleep(2)` (AFAICT that is equivalent to blocking the execution of the test for 2s while `call.remote` executes in a separate worker process) after `ray.wait` in the diff above and test passed. 
6. The behaviour is consistent across multiple runs of the test.
7. My above two comments in this thread also apply.

Reply:
Ah, thanks, that explanation clarifies the race condition. `requests.get` can take so long to execute that the line `assert len(ray.wait([ref2], timeout=0.1)[0]) == 0` passes, but this can cause `requests.get` to issue a request to the `V2` deployment instead of the `V1` deployment.

I want to say that bumping the timeout should work for now, but you're right that this still leaves a possible race condition. @pcmoritz @simon-mo What are your thoughts?

Reply:
This sounds good to me bumping timeout make sense. This is a race condition for testing only, not the actual Serve functionality. Nive investigation!

Reply:
Yes. It all depends on how fast `requests.get` executes on an OS. If it's slow enough then there will be a race condition otherwise everything will be fine. May be we can do something more robust. Like, instead of a timeout bump we can make a non-blocking `requests.get` call (in the test code just after this `ray.wait` line) to the same location which `call.remote` is trying to access. That way we will know that if `call.remote` accessed the right server or not. Anyways, if timeout works then complicating things might not be a good idea?